pkg/serving/config_changes.go

```diff
@@ -193,6 +194,24 @@ func UpdateAutoscaleWindow(template *servingv1alpha1.RevisionTemplateSpec, windo
 	return UpdateRevisionTemplateAnnotation(template, autoscaling.WindowAnnotationKey, window)
 }
 
+// UpdateClusterLocal updates container serving.knative.dev/visibility annotation
+func UpdateClusterLocal(service *servingv1alpha1.Service, template *servingv1alpha1.RevisionTemplateSpec, clusterLocal bool) error {
+	currentClusterLocal := template.Labels[config.VisibilityLabelKey]
+	if clusterLocal {
+		// NOOP when setting --cluster-local on already private service
+		if currentClusterLocal == config.VisibilityClusterLocal {
+			return nil
```

Comment:
even though this takes care of verifying NOOP change in config, 

`AnyMutation`  check sees the ConfigurationEditFlags has changed (`cluster-local`) and generates a new revision name for template. The new revision generated is because of new value for revision name.

https://github.com/knative/client/blob/ffdeafe33e29429830499005f2adbd4ccd8c7476/pkg/kn/commands/service/configuration_edit_flags.go#L253-L254

Reply:
Yup. Saw that last night too.

Not clear how to fix this since `AnyMutation` happens early in `Update` and checks for any change of any of the flags.

One possibility is to have a list of non-mutable flags + conditions function. If the flag is mutated and the condition function is false then no mutation happens.

I can attempt this unless I hear a better approach?

cc @rhuss

Reply:
Actually scrap above ^^^ the easier thing to do is to no mark as making revision:

```diff
+-	p.markFlagMakesRevision("cluster-local")
```

Reply:
> Actually scrap above ^^^ the easier thing to do is to no mark as making revision:

then it'd fail to generate revision when it has to

```
➜  client git:(b96455fd) ✗ ./kn service list
NAME   URL                              LATEST        AGE   CONDITIONS   READY   REASON
svc    http://svc.default.example.com   svc-xvcsh-1   30s   3 OK / 3     True    
➜  client git:(b96455fd) ✗ ./kn service update svc --cluster-local
admission webhook "validation.webhook.serving.knative.dev" denied the request: validation failed: Saw the following changes without a name change (-old +new): spec.template.metadata.name
{*v1.RevisionTemplateSpec}.ObjectMeta.Labels:
	-: "map[]"
	+: "map[serving.knative.dev/visibility:cluster-local]"
```
This approach won't work.

This issue is with every other flag which markes the revision, so we'd need to solve this for other flags as well.
We'd need to extend the check for `AnyMutation` in a way which not only rely on whether flags have `Changed` but also see if their values changed.
I will try and fix this another PR.

@maximilien : Please revert the latest change you made (as mentioned in your last comment) and add e2e tests please! (e2e test could've caught the error I pasted above with webhook validation)

Reply:
filed https://github.com/knative/client/issues/646

Reply:
OK I will add the `e2e` but I think the solution might be to have a:

`p.markFlagPotentiallyMakesRevision("cluster-local", func(flag, service, revision){...} bool)`

Where `func(flag, service, revision){...} bool` is a custom function for the flag that `AnyMutation` can call to check if that flag change does result in mutation?

I would like to attempt above but I think it should be done separately to address #646

Reply:
OK @navidshaikh added e2e basic private tests. When we have solution for #646 I can add for the cases of private/private and public/public updates which will fail.

Let me know on above on whether you are addressing #646 or shall I?